dipy/reconst/dki.py

```diff
@@ -2336,6 +2349,255 @@ def wls_fit_dki(design_matrix, data):
+            msg += " One of %s" % ', '.join(cvxpy.installed_solvers())
+            msg += " was expected."
+            raise ValueError(msg)
+
+    # Set up SDP solver
+    sdp_constraints = load_sdp_constraints('dki')
+    sdp = PositiveDefiniteLeastSquares(22, A=sdp_constraints)
+
+    # preparing data and initializing parameters
+    data = np.asarray(data)
+    data_flat = data.reshape((-1, data.shape[-1]))
+    dki_params = np.empty((len(data_flat), 27))
+
+    # looping OLS solution on all data voxels
+    for vox in range(len(data_flat)):
+        dki_params[vox] = _cls_iter(sdp, design_matrix, data_flat[vox],
+                                    cvxpy_solver=cvxpy_solver)
+
+    # Reshape data according to the input data shape
+    dki_params = dki_params.reshape((data.shape[:-1]) + (27,))
+
+    return dki_params
+
+
+def _cwls_iter(sdp, design_matrix, inv_design, sig,
+               cvxpy_solver=None):
+    """ Helper function used by cwls_fit_dki - Applies an LMI constrained
+    WLS fit of the diffusion kurtosis model to single voxel signals.
+
+    Parameters
+    ----------
+    sdp : PositiveDefiniteLeastSquares instance
+        A CVXPY representation of a regularized least squares optimization
+        problem.
+    design_matrix : array (g, 22)
+        Design matrix holding the covariants used to solve for the regression
+        coefficients.
+    inv_design : array (g, 22)
+        Inverse of the design matrix.
+    sig : array (g,)
+        Diffusion-weighted signal for a single voxel data.
+    cvxpy_solver : str, optional
+        cvxpy solver name. Optionally optimize the positivity constraint with a
+        particular cvxpy solver. See http://www.cvxpy.org/ for details.
+        Default: None (cvxpy chooses its own solver).
+
+    Returns
+    -------
+    dki_params : array (27,)
+        All parameters estimated from the diffusion kurtosis model.
+        Parameters are ordered as follows:
+            1) Three diffusion tensor's eigenvalues
+            2) Three lines of the eigenvector matrix each containing the first,
+               second and third coordinates of the eigenvector
+            3) Fifteen elements of the kurtosis tensor
+
+    """
+
+    # DKI ordinary linear least square solution (initial guess)
+    log_s = np.log(sig)
+    ols_result = np.dot(inv_design, log_s)
+
+    # Define sqrt weights as diag(yn)
+    W = np.diag(np.exp(np.dot(design_matrix, ols_result)))
+
+    # Set up sdp
+    result = sdp.solve(np.dot(W, design_matrix), np.dot(W, log_s), check=True,
+                       solver=cvxpy_solver)
+
+    # Extracting the diffusion tensor parameters from solution
+    DT_elements = result[:6]
+    evals, evecs = decompose_tensor(from_lower_triangular(DT_elements))
+
+    # Extracting kurtosis tensor parameters from solution
+    MD_square = (evals.mean(0))**2
+    if MD_square:
+        KT_elements = result[6:21] / MD_square
+    else:
+        KT_elements = 0.*result[6:21]
```

Comment:
> Should I rewrite those as well then?

if you have time, yes! Maybe @arokem or @RafaelNH have another opinion concerning this point. 

Otherwise, you can just focus to reduce the duplication in your code. 

> would it make sense to adapt the code so that it uses the multi_voxel_fit decorator?

oh yes! it would make a lot of sense and prepare for the future `parallel_voxel_fit decorator`. 

> Since the constrained fits take a while it is nice to be able to have a progress monitor like that provided by the decorator.

Completely agree with that